### PR TITLE
Internal tracking of InstanceHandles

### PR DESCRIPTION
**Pull request by mitza-oci**
- DomainParticipantImpl::get_handle previously could do two different things, these are now separated into assign_handle and lookup_handle
- Added DomainParticipantImpl::return_handle

### CODE REVIEW

File: dds/DCPS/DomainParticipantImpl.cpp [hunk 1]
```diff
@@ -1910,49 +1887,70 @@ DomainParticipantImpl::get_unique_id()
...
+    const DDS::InstanceHandle_t handle = participant_handles_.next();
+    ACE_DEBUG((LM_DEBUG, "New mapped IH %d for %C\n", handle, LogGuid(id).c_str()));
+    handles_[id] = std::make_pair(handle, 1);
+    repoIds_[handle] = id;
+    return handle;
   }
 
-  return result;
+  HandleWithCounter& mapped = location->second;
+  ++mapped.second;
```
[the remark this answers is not in the document]
Comment (mitza-oci) [reply dds/DCPS/DomainParticipantImpl.cpp:1913]: If the same GUID is used for N calls to `assign_handle`, there will be N calls to `return_handle`.

**Review by jrw972 — APPROVED** [review]

File: dds/DCPS/DataReaderImpl_T.h [hunk 2]
```diff
@@ -1733,16 +1733,17 @@ void store_instance_data(unique_ptr<MessageTypeWithAllocator> instance_data,
 
     just_registered = true;
     DDS::BuiltinTopicKey_t key = OpenDDS::DCPS::keyFromSample(static_cast<MessageType*>(instance_data.get()));
-    handle = handle == DDS::HANDLE_NIL ? this->get_next_handle( key) : handle;
+    bool owns_handle = false;
+    if (handle == DDS::HANDLE_NIL) {
+      handle = this->get_next_handle(key);
```
Comment (simpsont-oci) [inline dds/DCPS/DataReaderImpl_T.h:1738]:
```suggestion
      handle = get_next_handle(key);
```
ditto everywhere

File: dds/DCPS/DomainParticipantImpl.h [hunk 3]
```diff
@@ -476,7 +500,7 @@ class OpenDDS_Dcps_Export DomainParticipantImpl
   /// Protect the topic collection.
   ACE_Recursive_Thread_Mutex topics_protector_;
   /// Protect the handle collection.
-  ACE_Recursive_Thread_Mutex handle_protector_;
+  mutable ACE_Thread_Mutex handle_protector_;
```
Comment (simpsont-oci) [inline dds/DCPS/DomainParticipantImpl.h:503]: The atomic coordination between two different data structures makes me think this ought to live in its own utility class and have a small unit test.

File: dds/DCPS/SubscriptionInstance.h [hunk 4]
```diff
@@ -81,6 +58,8 @@ class SubscriptionInstance : public RcObject {
   /// The instance handle for the registered object
   DDS::InstanceHandle_t instance_handle_;
 
+  bool owns_handle_;
```
Comment (simpsont-oci) [inline dds/DCPS/SubscriptionInstance.h:61]: const?

File: dds/DCPS/DataReaderImpl.cpp [hunk 5]
```diff
@@ -375,15 +372,15 @@ DataReaderImpl::transport_assoc_done(int flags, const RepoId& remote_id)
   }
   // We no longer hold the publication_handle_lock_.
 
-  if (!is_bit_) {
 
-    RcHandle<DomainParticipantImpl> participant = this->participant_servant_.lock();
+  RcHandle<DomainParticipantImpl> participant = this->participant_servant_.lock();
```
[the remark this answers is not in the document]
Comment (mitza-oci) [reply dds/DCPS/DataReaderImpl.cpp:376]: Right, I didn't update all of these.  Probably good to go ahead with it in this PR.

File: dds/DCPS/DomainParticipantImpl.h (hunk 3 above)
Comment (mitza-oci) [reply dds/DCPS/DomainParticipantImpl.h:503]: Yes, but we may not be able to do that in this PR.  The pattern is essentially https://www.boost.org/doc/libs/1_66_0/libs/bimap/doc/html

**Review by simpsont-oci — APPROVED** [review]

**Pull request merged**